packages/discovery-provider/src/queries/comments/utils.py

```diff
@@ -0,0 +1,446 @@
+
+logger = logging.getLogger(__name__)
+
+# Constants
+COMMENT_ROOT_DEFAULT_LIMIT = 15  # default pagination limit
+COMMENT_REPLIES_DEFAULT_LIMIT = 3  # default replies pagination limit
+
+
+# Returns whether a comment has been reacted to by a particular user
+def get_is_reacted(session, user_id, comment_id):
+    if not user_id:
+        return False
+    is_react = (
+        session.query(CommentReaction)
+        .filter(
+            CommentReaction.user_id == user_id,
+            CommentReaction.comment_id == comment_id,
+            CommentReaction.is_delete == False,
+        )
+        .first()
+    )
+    return is_react is not None
+
+
+def get_base_comments_query(session, current_user_id=None):
+    """Base query builder for comments with common functionality"""
+    mentioned_users = (
+        session.query(
+            CommentMention.comment_id,
+            CommentMention.user_id,
+            User.handle,
+            CommentMention.is_delete,
+        )
+        .join(User, CommentMention.user_id == User.user_id)
+        .subquery()
+    )
+
+    react_count_subquery = (
+        session.query(
+            CommentReaction.comment_id,
+            func.count(CommentReaction.comment_id).label("react_count"),
+        )
+        .filter(CommentReaction.is_delete == False)
+        .group_by(CommentReaction.comment_id)
+        .subquery()
+    )
+
+    muted_by_karma = (
+        session.query(MutedUser.muted_user_id)
+        .join(AggregateUser, MutedUser.user_id == AggregateUser.user_id)
+        .filter(MutedUser.is_delete == False)
+        .group_by(MutedUser.muted_user_id)
+        .having(func.sum(AggregateUser.follower_count) >= COMMENT_KARMA_THRESHOLD)
+        .subquery()
+    )
+
+    return {
+        "mentioned_users": mentioned_users,
+        "react_count_subquery": react_count_subquery,
+        "muted_by_karma": muted_by_karma,
+    }
+
+
+def _format_comment_response(
+    comment,
+    react_count,
+    is_muted,
+    mentions,
+    current_user_id,
+    session,
+    include_replies=False,
+    artist_id=None,
+):
+    """Common formatter for comment responses"""
+
+    def remove_delete(mention):
+        del mention["is_delete"]
+        return mention
+
+    def filter_mentions(mention):
+        return mention["user_id"] is not None and mention["is_delete"] is not True
+
+    response = {
+        "id": encode_int_id(comment.comment_id),
+        "entity_id": encode_int_id(comment.entity_id),
+        "entity_type": comment.entity_type,
+        "user_id": encode_int_id(comment.user_id) if not comment.is_delete else None,
+        "mentions": list(map(remove_delete, filter(filter_mentions, mentions))),
+        "message": comment.text if not comment.is_delete else "[Removed]",
+        "is_edited": comment.is_edited,
+        "track_timestamp_s": comment.track_timestamp_s,
+        "react_count": react_count,
+        "is_current_user_reacted": get_is_reacted(
```

Comment:
Reading this now I'm realizing we execute a separate query for every comment we are processing?
We could/should look at doing this in bulk the way we do for other entities when merging in the contextual user data.

Reply:
100% will fix note and fix this